File: android/app/src/main/java/com/liive/ride/designsystem/LiiveButton.kt
//  LiiveButton.kt
//  Liive Ride — sample component showing how DS tokens compose in Compose.
//  Mirror this pattern for Badge, Card, ListRow, GlassPanel, SosButton, etc.
package com.liive.ride.designsystem

import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.annotation.DrawableRes
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow

enum class LiiveButtonVariant { Primary, Secondary, Tinted, Plain, Destructive, DestructivePlain }
enum class LiiveButtonSize { Sm, Md, Lg }

@Composable
fun LiiveButton(
    title: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    variant: LiiveButtonVariant = LiiveButtonVariant.Primary,
    size: LiiveButtonSize = LiiveButtonSize.Md,
    capsule: Boolean = false,
    fullWidth: Boolean = false,
    tabularNumbers: Boolean = false,
    @DrawableRes icon: Int? = null,
    @DrawableRes iconRight: Int? = null,
    iconOnly: Boolean = false,
    contentDescription: String? = null,
    disabled: Boolean = false,
    loading: Boolean = false,
) {
    val c = LiiveTheme.colors
    val interaction = remember { MutableInteractionSource() }
    val pressed by interaction.collectIsPressedAsState()
    val enabled = !disabled && !loading

    val height = when (size) {
        LiiveButtonSize.Sm -> LiiveButtonLayout.SmallHeight
        LiiveButtonSize.Md -> LiiveButtonLayout.MediumHeight
        LiiveButtonSize.Lg -> LiiveButtonLayout.LargeHeight
    }
    val horizontalPadding = when (size) {
        LiiveButtonSize.Sm -> LiiveButtonLayout.SmallHorizontalPadding
        LiiveButtonSize.Md -> LiiveButtonLayout.MediumHorizontalPadding
        LiiveButtonSize.Lg -> LiiveButtonLayout.LargeHorizontalPadding
    }
    val bgPressed = when (variant) {
        LiiveButtonVariant.Primary -> c.accentPressed
        LiiveButtonVariant.Secondary -> c.fillSecondary
        LiiveButtonVariant.Tinted -> c.accentTint
        LiiveButtonVariant.Plain, LiiveButtonVariant.DestructivePlain -> LiiveButtonLayout.TransparentColor
        LiiveButtonVariant.Destructive -> c.danger
    }
    val bg = if (enabled && pressed) {
        bgPressed
    } else {
        when (variant) {
            LiiveButtonVariant.Primary -> c.accent
            LiiveButtonVariant.Secondary -> c.fill
            LiiveButtonVariant.Tinted -> c.accentTint
            LiiveButtonVariant.Plain, LiiveButtonVariant.DestructivePlain -> LiiveButtonLayout.TransparentColor
            LiiveButtonVariant.Destructive -> c.danger
        }
    }
    val fg = when (variant) {
        LiiveButtonVariant.Primary -> c.onAccent
        LiiveButtonVariant.Secondary -> c.text
        LiiveButtonVariant.Tinted, LiiveButtonVariant.Plain -> c.accent
        LiiveButtonVariant.Destructive -> LiiveButtonLayout.DestructiveForegroundColor
        LiiveButtonVariant.DestructivePlain -> c.danger
    }
    val labelWeight = when (variant) {
        LiiveButtonVariant.Plain, LiiveButtonVariant.DestructivePlain -> FontWeight.Normal
        else -> FontWeight.SemiBold
    }
    val labelStyle = (if (size == LiiveButtonSize.Sm) {
        MaterialTheme.typography.titleMedium
    } else {
        MaterialTheme.typography.titleLarge
    }).copy(
        fontWeight = labelWeight,
        letterSpacing = MaterialTheme.typography.headlineSmall.letterSpacing
    ).let { style ->
        if (tabularNumbers) style.tabularNumbers() else style
    }
    val shape: RoundedCornerShape = if (capsule) LiiveRadius.full else LiiveRadius.md
    val targetOpacity = when {
        disabled -> LiiveButtonLayout.DisabledOpacity
        enabled && pressed && variant in setOf(
            LiiveButtonVariant.Plain,
            LiiveButtonVariant.Tinted,
            LiiveButtonVariant.DestructivePlain
        ) -> LiiveButtonLayout.SubtlePressedOpacity
        enabled && pressed -> LiiveButtonLayout.FilledPressedOpacity
        else -> LiiveButtonLayout.EnabledOpacity
    }
    val motionSpec = tween<Float>(durationMillis = LiiveMotion.fastMs, easing = LiiveMotion.easeOut)
    val animatedScale by animateFloatAsState(
        targetValue = if (enabled && pressed) LiiveMotion.pressScale else LiiveButtonLayout.RestingScale,
        animationSpec = motionSpec,
        label = "LiiveButtonScale"
    )
    val animatedOpacity by animateFloatAsState(
        targetValue = targetOpacity,
        animationSpec = motionSpec,
        label = "LiiveButtonOpacity"
    )
    val animatedBackground by animateColorAsState(
        targetValue = bg,
        animationSpec = tween(durationMillis = LiiveMotion.fastMs, easing = LiiveMotion.easeOut),
        label = "LiiveButtonBackground"
    )

    Box(
        modifier = modifier
            .then(
                when {
                    fullWidth -> Modifier.fillMaxWidth()
                    iconOnly -> Modifier.width(height)
                    else -> Modifier
                }
            )
            .height(height)
            .scale(animatedScale)
            .alpha(animatedOpacity)
            .clip(shape)
            .background(animatedBackground)
            .clickable(interactionSource = interaction, indication = null, enabled = enabled) { onClick() }
            .padding(horizontal = if (iconOnly) LiiveButtonLayout.IconOnlyHorizontalPadding else horizontalPadding),
        contentAlignment = Alignment.Center,
    ) {
        if (loading) {
            CircularProgressIndicator(
                color = fg,
                strokeWidth = LiiveButtonLayout.SpinnerStrokeWidth,
                modifier = Modifier.size(LiiveButtonLayout.SpinnerSize)
            )
        } else if (iconOnly && icon != null) {
            Icon(
                painterResource(icon),
                contentDescription,
                tint = fg,
                modifier = Modifier.size(LiiveButtonLayout.IconSize)
            )
        } else {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(LiiveButtonLayout.ContentGap)
            ) {
                if (icon != null) {
                    Icon(
                        painterResource(icon),
                        null,
                        tint = fg,
                        modifier = Modifier.size(LiiveButtonLayout.IconSize)
                    )
                }
                Text(
                    text = title,
                    color = fg,
                    style = labelStyle,
                    textAlign = TextAlign.Center,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
                if (iconRight != null) {
                    Icon(
                        painterResource(iconRight),
                        null,
                        tint = fg,
                        modifier = Modifier.size(LiiveButtonLayout.IconSize)
                    )
                }
            }
        }
    }
}

private object LiiveButtonLayout {
    val SmallHeight = LiiveControl.sm
    val MediumHeight = LiiveControl.md
    val LargeHeight = LiiveControl.lg
    val SmallHorizontalPadding = LiiveSpacing.m + LiiveSpacing.xs2
    val MediumHorizontalPadding = LiiveSpacing.l + LiiveSpacing.xs2
    val LargeHorizontalPadding = LiiveSpacing.xxl - LiiveSpacing.xs2
    val IconOnlyHorizontalPadding = LiiveSpacing.xs2 - LiiveSpacing.xs2
    val ContentGap = LiiveSpacing.s
    val IconSize = LiiveSpacing.l + LiiveSpacing.xs2
    val SpinnerSize = IconSize
    val SpinnerStrokeWidth = LiiveSpacing.xs2
    const val DisabledOpacity = 0.4f
    const val SubtlePressedOpacity = 0.5f
    const val FilledPressedOpacity = 0.85f
    const val EnabledOpacity = 1f
    const val RestingScale = 1f
    val TransparentColor = Color.Transparent
    val DestructiveForegroundColor = Color.White
}

File: Packages/RideSharingFeature/Sources/RideSharingFeature/DesignSystem/LiiveButton.swift
//  LiiveButton.swift
//  Liive Ride — sample component showing how DS tokens compose in SwiftUI.
//  Mirror this pattern for Badge, Card, ListRow, GlassPanel, SOSButton, etc.

import SwiftUI

public struct LiiveButton: View {
    public enum Variant { case primary, secondary, tinted, plain, destructive, destructivePlain }
    public enum Size { case sm, md, lg }
    public enum Shape { case rounded, capsule }

    let title: String
    var variant: Variant = .primary
    var size: Size = .md
    var shape: Shape = .rounded
    var fullWidth: Bool = false
    var icon: Image? = nil
    var iconRight: Image? = nil
    var iconOnly: Bool = false
    var accessibilityLabel: String? = nil
    var tabularNumbers: Bool = false
    var disabled: Bool = false
    var loading: Bool = false
    let action: () -> Void

    @GestureState private var pressed = false

    public init(_ title: String, variant: Variant = .primary, size: Size = .md,
                shape: Shape = .rounded, fullWidth: Bool = false, icon: Image? = nil,
                iconRight: Image? = nil, iconOnly: Bool = false, accessibilityLabel: String? = nil,
                tabularNumbers: Bool = false, disabled: Bool = false, loading: Bool = false,
                action: @escaping () -> Void) {
        self.title = title; self.variant = variant; self.size = size
        self.shape = shape; self.fullWidth = fullWidth; self.icon = icon
        self.iconRight = iconRight; self.iconOnly = iconOnly
        self.accessibilityLabel = accessibilityLabel; self.tabularNumbers = tabularNumbers
        self.disabled = disabled; self.loading = loading; self.action = action
    }

    private var height: CGFloat {
        switch size {
        case .sm: return LiiveButtonLayout.smallHeight
        case .md: return LiiveButtonLayout.mediumHeight
        case .lg: return LiiveButtonLayout.largeHeight
        }
    }
    private var horizontalPadding: CGFloat {
        switch size {
        case .sm: return LiiveButtonLayout.smallHorizontalPadding
        case .md: return LiiveButtonLayout.mediumHorizontalPadding
        case .lg: return LiiveButtonLayout.largeHorizontalPadding
        }
    }
    private var titleFont: Font {
        size == .sm ? LiiveFont.subhead : LiiveFont.headline
    }
    private var titleWeight: Font.Weight {
        switch variant {
        case .plain, .destructivePlain: return .regular
        default: return .semibold
        }
    }
    private var bg: Color {
        if pressed { return bgPressed }
        switch variant {
        case .primary: return LiiveColor.accent
        case .secondary: return LiiveColor.fill
        case .tinted: return LiiveColor.accentTint
        case .plain, .destructivePlain: return .clear
        case .destructive: return LiiveColor.danger
        }
    }
    private var bgPressed: Color {
        switch variant {
        case .primary: return LiiveColor.accentPressed
        case .secondary: return LiiveColor.fillSecondary
        case .tinted: return LiiveColor.accentTint
        case .plain, .destructivePlain: return .clear
        case .destructive: return LiiveColor.danger
        }
    }
    private var fg: Color {
        switch variant {
        case .primary: return LiiveColor.onAccent
        case .secondary: return LiiveColor.text
        case .tinted: return LiiveColor.accent
        case .plain: return LiiveColor.accent
        case .destructive: return LiiveButtonLayout.destructiveForegroundColor
        case .destructivePlain: return LiiveColor.danger
        }
    }
    private var radius: CGFloat { shape == .capsule ? LiiveRadius.full : LiiveRadius.md }
    private var isInteractive: Bool { !disabled && !loading }
    private var opacity: Double {
        if disabled { return LiiveButtonLayout.disabledOpacity }
        if pressed {
            switch variant {
            case .plain, .tinted, .destructivePlain: return LiiveButtonLayout.subtlePressedOpacity
            default: return LiiveButtonLayout.filledPressedOpacity
            }
        }
        return LiiveButtonLayout.enabledOpacity
    }

    public var body: some View {
        Button(action: { if isInteractive { action() } }) {
            HStack(spacing: LiiveButtonLayout.contentGap) {
                if loading {
                    ProgressView()
                        .progressViewStyle(.circular)
                        .tint(fg)
                        .frame(width: LiiveButtonLayout.spinnerSize, height: LiiveButtonLayout.spinnerSize)
                } else if iconOnly, let icon {
                    icon
                        .font(.system(size: LiiveButtonLayout.iconSize, weight: .semibold))
                } else {
                    if let icon { icon }
                    Text(title)
                        .font((tabularNumbers ? titleFont.monospacedDigit() : titleFont).weight(titleWeight))
                        .tracking(LiiveFont.Tracking.title3)
                    if let iconRight { iconRight }
                }
            }
            .frame(maxWidth: fullWidth ? .infinity : nil)
            .frame(width: iconOnly && !fullWidth ? height : nil)
            .frame(height: height)
            .padding(.horizontal, iconOnly ? LiiveButtonLayout.iconOnlyHorizontalPadding : horizontalPadding)
            .foregroundColor(fg)
            .background(bg)
            .clipShape(RoundedRectangle(cornerRadius: radius, style: .continuous))
            .scaleEffect(pressed ? LiiveMotion.pressScale : LiiveButtonLayout.restingScale)
            .opacity(opacity)
            .animation(.easeOut(duration: LiiveMotion.fast), value: pressed)
        }
        .buttonStyle(.plain)
        .disabled(disabled || loading)
        .accessibilityLabel(Text(accessibilityLabel ?? title))
        .simultaneousGesture(
            DragGesture(minimumDistance: LiiveButtonLayout.dragMinimumDistance)
                .updating($pressed) { _, state, _ in
                    state = isInteractive
                }
        )
    }
}

private enum LiiveButtonLayout {
    static let smallHeight = LiiveControl.sm
    static let mediumHeight = LiiveControl.md
    static let largeHeight = LiiveControl.lg
    static let smallHorizontalPadding = LiiveSpacing.m + LiiveSpacing.xs2
    static let mediumHorizontalPadding = LiiveSpacing.l + LiiveSpacing.xs2
    static let largeHorizontalPadding = LiiveSpacing.xxl - LiiveSpacing.xs2
    static let iconOnlyHorizontalPadding = LiiveSpacing.xs2 - LiiveSpacing.xs2
    static let contentGap = LiiveSpacing.s
    static let iconSize = LiiveSpacing.l + LiiveSpacing.xs2
    static let spinnerSize = iconSize
    static let disabledOpacity = 0.4
    static let subtlePressedOpacity = 0.5
    static let filledPressedOpacity = 0.85
    static let enabledOpacity = 1.0
    static let restingScale: CGFloat = 1.0
    static let dragMinimumDistance = LiiveSpacing.xs2 - LiiveSpacing.xs2
    static let destructiveForegroundColor = Color.white
}
